File: android/core/src/main/kotlin/dev/mekholi/core/Sync.kt
package dev.mekholi.core

/**
 * The connection, as the platform sees it.
 *
 * A hint, never the truth: a captive portal reports "online" and a shop with a
 * dead uplink reports "online" too. It decides *when to try*, never whether a
 * sale is safe — the outbox's own results are the truth.
 */
interface Connectivity {
    fun isOnline(): Boolean
    /** Returns an unsubscribe function. Fires in both directions. */
    fun onChange(listener: (Boolean) -> Unit): () -> Unit
}

/** A one-shot timer, so nothing in this module needs an Android `Handler`. */
interface Scheduler {
    fun schedule(delayMs: Long, task: () -> Unit): () -> Unit
}

data class SyncStatus(
    val online: Boolean = true,
    val pending: Int = 0,
    val failed: Int = 0,
    val syncing: Boolean = false,
    val lastSyncedAt: Long? = null,
    val lastError: String? = null,
) {
    /** True when nothing is left to send, and nothing a person must decide about. */
    val settled: Boolean get() = pending == 0 && failed == 0
}

/**
 * Sends what the outbox is holding, and says what it is holding.
 *
 * Same shape and same guarantees as `SyncEngine` in the web client
 * (`src/shared/repositories/offline/sync.ts`), because the two are answering
 * the same question and a shop that switched between them must not be able to
 * tell which one it was using.
 */
class SyncEngine(
    private val api: MekholiApi,
    private val outbox: Outbox,
    private val connectivity: Connectivity,
    /** The live token: a till is signed in for hours, and tokens expire. */
    private val token: () -> String?,
    private val onStatus: (SyncStatus) -> Unit = {},
    private val scheduler: Scheduler? = null,
    private val intervalMs: Long = 20_000,
    private val now: () -> Long = { System.currentTimeMillis() },
) {

    private var unsubscribe: (() -> Unit)? = null
    private var cancelTimer: (() -> Unit)? = null
    private var draining = false

    var status: SyncStatus = SyncStatus()
        private set

    fun start(): SyncStatus {
        if (unsubscribe == null) {
            unsubscribe = connectivity.onChange { online ->
                publish(status.copy(online = online))
                if (online) drain()
            }
        }
        refresh()
        if (status.online && status.pending > 0) drain()
        schedule()
        return status
    }

    fun stop() {
        unsubscribe?.invoke()
        unsubscribe = null
        cancelTimer?.invoke()
        cancelTimer = null
    }

    /** Recompute the counts from the outbox, and publish. */
    fun refresh(): SyncStatus {
        publish(
            status.copy(
                online = connectivity.isOnline(),
                pending = outbox.pending().size,
                failed = outbox.failures().size,
            )
        )
        return status
    }

    /**
     * Try now, whether or not the platform thinks there is a connection.
     *
     * Single-flight: a drain that is already running is returned rather than
     * started again, because two drains racing is how the same sale gets sent
     * twice — which the server would survive (§044) but which is still work
     * nobody asked for.
     */
    @Synchronized
    fun drain(): DrainResult {
        if (draining) return DrainResult(0, 0, true, outbox.size())
        draining = true
        publish(status.copy(syncing = true))

        val result = try {
            outbox.drain { send(it) }
        } finally {
            draining = false
            publish(status.copy(syncing = false))
            refresh()
            schedule()
        }

        if (result.sent > 0 && !result.stopped) {
            publish(status.copy(lastSyncedAt = now(), lastError = null))
        }
        return result
    }

    /** Put a refused sale back in line and try again. */
    fun retry(ref: String) {
        outbox.retry(ref)
        refresh()
        drain()
    }

    fun discard(ref: String): QueuedWrite? {
        val removed = outbox.discard(ref)
        refresh()
        return removed
    }

    fun failures(): List<QueuedWrite> = outbox.failures()

    /**
     * Keep a sale the connection would not take, and hand back its reference.
     *
     * The body is the exact JSON of the RPC call — minted by `MekholiApi.encodeSale`
     * from the request the till built — so the replay is the same request with
     * the same reference, which is the whole of the exactly-once story.
     */
    fun enqueueSale(body: String, ref: String): QueuedWrite =
        outbox.enqueue(KIND_COMPLETE_SALE, body, ref)

    /**
     * One send of one queued sale.
     *
     * `23505` is the unique index on `(organization_id, client_ref)` saying
     * another attempt already wrote this sale — that is the outcome the outbox
     * wanted, so it counts as success. Treating it as a refusal would leave the
     * shop chasing a sale it already has.
     */
    private fun send(write: QueuedWrite): SendOutcome {
        val accessToken = token()
            ?: return SendOutcome(false, FailureKind.OFFLINE, "Not signed in.")
        return try {
            if (write.kind == KIND_COMPLETE_SALE) {
                api.completeSale(accessToken, write.body)
            }
            SendOutcome(true)
        } catch (error: ApiException) {
            if (error.code == DUPLICATE_KEY) SendOutcome(true)
            else SendOutcome(false, FailureKind.REFUSED, describe(error))
        } catch (error: Throwable) {
            val kind = classify(error)
            SendOutcome(false, kind, describe(error))
        }
    }

    private fun publish(next: SyncStatus) {
        val was = status
        status = next
        if (was != next) onStatus(next)
    }

    /**
     * One timer, re-armed rather than repeated. A `setInterval` racing a slow
     * upload is how a queue sends the same sale twice.
     */
    private fun schedule() {
        val timers = scheduler ?: return
        cancelTimer?.invoke()
        cancelTimer = null
        if (status.pending == 0 || !status.online) return
        cancelTimer = timers.schedule(intervalMs) { drain() }
    }

    companion object {
        const val KIND_COMPLETE_SALE = "sale.complete"
        const val DUPLICATE_KEY = "23505"
    }
}

File: src/shared/repositories/offline/sync.ts
/**
 * The sync engine: what happens between "the wifi came back" and "the till is
 * up to date".
 *
 * It owns three decisions, and nothing else:
 *
 *   · *When to try.* `navigator.onLine` is a hint, not a fact — a captive
 *     portal, a dead uplink and a sleeping laptop all report `true`. So the
 *     flag only ever speeds things up; the queue's own answers decide whether
 *     the shop is really connected, and a manual retry always tries.
 *
 *   · *What a failure means.* Handled by the injected classifier (see
 *     `queue.ts`): the network is "try later", an answer from the server is
 *     "somebody has to decide". This file never guesses.
 *
 *   · *What the cashier is told.* A status object with counts and the last
 *     error, published on every change so the shell can render it. Silence is
 *     the failure mode to avoid: a queued sale nobody mentions is a sale
 *     somebody will re-take by hand.
 */

import type { DrainResult, QueuedWrite, SendOutcome } from './queue'
import type { WriteQueue } from './queue'

export interface SyncStatus {
  /** The browser's opinion, which is why it is labelled as such. */
  online: boolean
  pending: number
  failed: number
  syncing: boolean
  /** Epoch ms of the last successful drain, or null if none has happened. */
  lastSyncedAt: number | null
  lastError: string | null
  /** False when the store could not survive a reload (private mode). */
  persistent: boolean
}

/** The last attempt's outcome, per write — what the failure list shows. */
export interface SyncFailure {
  ref: string
  kind: string
  createdAt: number
  attempts: number
  message: string
}

export interface Connectivity {
  isOnline(): boolean
  /** Returns an unsubscribe function. Fires on both directions. */
  onChange(listener: (online: boolean) => void): () => void
}

export interface SyncEngineOptions {
  queue: WriteQueue
  /** Performs the real call for one queued write. Injected: see `createSaleSender`. */
  send: (write: QueuedWrite) => Promise<SendOutcome>
  connectivity: Connectivity
  onStatus?: (status: SyncStatus) => void
  now?: () => number
  /** How often to try while there is something waiting and the flag says online. */
  intervalMs?: number
  /** Timer hooks, so a test drives the clock instead of waiting for it. */
  setTimer?: (fn: () => void, ms: number) => number
  clearTimer?: (handle: number) => void
  persistent?: boolean
}

export class SyncEngine {
  readonly #options: SyncEngineOptions
  readonly #now: () => number
  #status: SyncStatus
  #unsubscribe: (() => void) | null = null
  #timer: number | null = null
  #draining: Promise<DrainResult> | null = null

  constructor(options: SyncEngineOptions) {
    this.#options = options
    this.#now = options.now ?? (() => Date.now())
    this.#status = {
      online: options.connectivity.isOnline(),
      pending: 0,
      failed: 0,
      syncing: false,
      lastSyncedAt: null,
      lastError: null,
      persistent: options.persistent ?? true,
    }
  }

  get status(): SyncStatus {
    return this.#status
  }

  /** Attach to the browser's connectivity events and read the queue once. */
  async start(): Promise<SyncStatus> {
    if (this.#unsubscribe) return this.#status

    this.#unsubscribe = this.#options.connectivity.onChange((online) => {
      this.#publish({ online })
      if (online) void this.drain()
    })

    await this.refresh()
    if (this.#status.online && this.#status.pending > 0) void this.drain()
    this.#schedule()
    return this.#status
  }

  stop(): void {
    this.#unsubscribe?.()
    this.#unsubscribe = null
    if (this.#timer !== null && this.#options.clearTimer) this.#options.clearTimer(this.#timer)
    this.#timer = null
  }

  /** Recompute the counts from the queue, and publish. */
  async refresh(): Promise<SyncStatus> {
    const [pending, failed] = await Promise.all([
      this.#options.queue.pending(),
      this.#options.queue.failures(),
    ])
    this.#publish({ pending: pending.length, failed: failed.length })
    return this.#status
  }

  /** The failed writes, for the panel that lets a shopkeeper resolve them. */
  async failures(): Promise<SyncFailure[]> {
    const writes = await this.#options.queue.failures()
    return writes.map((write) => ({
      ref: write.ref,
      kind: write.kind,
      createdAt: write.createdAt,
      attempts: write.attempts,
      message: write.lastError ?? 'The server refused this sale.',
    }))
  }

  /** Try now, whether or not the browser thinks there is a connection. */
  async drain(): Promise<DrainResult> {
    if (this.#draining) return this.#draining

    this.#draining = (async () => {
      this.#publish({ syncing: true })
      try {
        const result = await this.#options.queue.drain(this.#options.send)
        if (result.sent > 0 && !result.stopped) {
          this.#publish({ lastSyncedAt: this.#now(), lastError: null })
        }
        return result
      } finally {
        this.#draining = null
        this.#publish({ syncing: false })
        await this.refresh()
        this.#schedule()
      }
    })()

    return this.#draining
  }

  /** Put a refused write back in line and try again. A human decided this. */
  async retry(ref: string): Promise<void> {
    await this.#options.queue.retry(ref)
    await this.refresh()
    await this.drain()
  }

  /** Throw a queued sale away. Only ever from an explicit choice. */
  async discard(ref: string): Promise<QueuedWrite | null> {
    const removed = await this.#options.queue.discard(ref)
    await this.refresh()
    return removed
  }

  /** Tell the engine something was just queued — try now if it can. */
  notifyQueued(): void {
    if (this.#status.online) void this.drain()
  }

  #publish(patch: Partial<SyncStatus>): void {
    const next: SyncStatus = { ...this.#status, ...patch }
    const changed = (Object.keys(next) as (keyof SyncStatus)[]).some((key) => next[key] !== this.#status[key])
    if (!changed) return
    this.#status = next
    this.#options.onStatus?.(next)
  }

  /**
   * One timer, re-armed rather than repeated: a drain that is still running
   * when the timer fires changes nothing, and a `setInterval` racing a slow
   * upload is how a queue sends the same sale twice.
   */
  #schedule(): void {
    const { setTimer, clearTimer, intervalMs } = this.#options
    if (!setTimer || !clearTimer) return
    if (this.#timer !== null) {
      clearTimer(this.#timer)
      this.#timer = null
    }
    if (this.#status.pending === 0 || !this.#status.online) return
    this.#timer = setTimer(() => {
      this.#timer = null
      void this.drain()
    }, intervalMs ?? 20_000)
  }
}

/**
 * The sender the engine should use for queued sales.
 *
 * Two things it gets right that a naive wrapper would not:
 *
 *   · `23505` is a **success**. It is the unique index on `(organization_id,
 *     client_ref)` saying another attempt already wrote this sale — exactly
 *     what happened, and exactly what the queue wanted. Treating it as a
 *     failure would leave the shop chasing a sale it already has.
 *   · The error is classified before it is returned, so "the network is gone"
 *     and "the shop refused this" take different paths through the queue.
 */
export function createSaleSender(
  complete: (payload: never) => Promise<unknown>,
  classify: (error: unknown) => 'offline' | 'refused',
  describe: (error: unknown) => string
): (write: QueuedWrite) => Promise<SendOutcome> {
  return async (write) => {
    try {
      await complete(write.payload as never)
      return { ok: true }
    } catch (error) {
      const code =
        typeof error === 'object' && error !== null ? (error as { code?: unknown }).code : undefined
      if (code === '23505') return { ok: true }
      const failure = classify(error)
      return { ok: false, failure, message: describe(error) }
    }
  }
}

/** The browser's connectivity, in the shape the engine wants. */
export function browserConnectivity(target: Window = window): Connectivity {
  return {
    isOnline: () => target.navigator.onLine,
    onChange(listener) {
      const onOnline = (): void => listener(true)
      const onOffline = (): void => listener(false)
      target.addEventListener('online', onOnline)
      target.addEventListener('offline', onOffline)
      return () => {
        target.removeEventListener('online', onOnline)
        target.removeEventListener('offline', onOffline)
      }
    },
  }
}
